nni/contrib/compression/quantization/qat_quantizer.py

```diff
@@ -0,0 +1,141 @@
+            if target_space.quant_scheme in ['symmetric', None]:
+                abs_max = torch.max(torch.abs(tracked_min), torch.abs(tracked_max))
+                scale = abs_max / (float(target_space.qmax - target_space.qmin) / 2)
+                scale = torch.max(scale, torch.full_like(scale, torch.finfo(torch.float32).eps))
+                # NOTE: here need to check, +1 because in pytorch, symmetric qint8 zp is 0, quint8 zp is 128.
+                zero_point_val = (target_space.qmax + target_space.qmin + 1) // 2
+                zero_point = zero_point.new_full(zero_point.size(), zero_point_val)
+            elif target_space.quant_scheme == 'affine':
+                scale = (tracked_max - tracked_min) / float(target_space.qmax - target_space.qmin)
+                scale = torch.max(scale, torch.full_like(scale, torch.finfo(torch.float32).eps))
+                zero_point = target_space.qmin - torch.round(tracked_min / scale)
+            else:
+                raise RuntimeError(f'Unknown quant_scheme {target_space.quant_scheme}')
+            zero_point = torch.clamp(zero_point, target_space.qmin, target_space.qmax)
+            target_space.scale, target_space.zero_point = scale, zero_point
+
+    def track_forward(self, *args, **kwargs):
+        super().track_forward(*args, **kwargs)
+        for _, ts in self._target_spaces.items():
+            for _, target_space in ts.items():
+                if target_space.scale is None:
+                    target_space.scale = torch.empty_like(target_space.tracked_max)
+                if target_space.zero_point is None:
+                    target_space.zero_point = torch.empty_like(target_space.tracked_max)
+
+    def register_trigger(self, evaluator: Evaluator):
+
+        def optimizer_task():
+            self.current_step += 1
+            if self.current_step == self.quant_start_step:
+                self.register_qat_apply_method()
+
+        evaluator.patch_optimizer_step(before_step_tasks=[], after_step_tasks=[optimizer_task])
+
+    def compress(self, max_steps: int | None = None, max_epochs: int | None = None):
+        self.evaluator.bind_model(self.bound_model, self._get_param_names_map())
+        self.register_trigger(self.evaluator)
+        self.evaluator.train(max_steps, max_epochs)
+        self.evaluator.unbind_model()
+        return self.bound_model, self.get_calibration_config()
+
+
+def track_min_max_val(wrapper: ModuleWrapper, target_name: str, target: Tensor):
+    def amin_reduce_func(converted_target: Tensor):
+        return converted_target.detach().amin(dim=-1)
+
+    def amax_reduce_func(converted_target: Tensor):
+        return converted_target.detach().amax(dim=-1)
+
+    if target_name in wrapper.quantization_target_spaces:
+        target_space = wrapper.quantization_target_spaces[target_name]
+        if target_space._scaler:
+            current_amin = target_space._scaler.shrink(target, amin_reduce_func, keepdim=True)
```

Comment:
if target_space._scaler == None, is there any problem? Should we provide a default value? here, if granularity ==  default, then target_space._scaler = None.